File: src/main/java/org/evochora/server/indexer/annotation/handlers/ParameterTokenHandler.java
package org.evochora.server.indexer.annotation.handlers;

import org.evochora.compiler.api.ProgramArtifact;
import org.evochora.compiler.api.TokenInfo;
import org.evochora.compiler.api.SourceInfo;
import org.evochora.runtime.isa.Instruction;
import org.evochora.server.contracts.raw.RawOrganismState;
import org.evochora.server.contracts.raw.SerializableProcFrame;
import org.evochora.server.indexer.annotation.ITokenHandler;
import org.evochora.server.indexer.annotation.TokenAnalysisResult;
import org.evochora.server.indexer.annotation.enums.TokenType;

import java.util.Deque;
import java.util.List;

/**
 * Handles tokens that are procedure parameter names.
 * Shows the bound register and its value: [%DRx=TYPE:VALUE] or [%DRx=x|y|...].
 * Works anywhere in procedure bodies, not just on .PROC directive lines.
 * Resolves parameter bindings by walking the call stack to find actual DR/PR registers.
 */
public class ParameterTokenHandler implements ITokenHandler {
    
    @Override
    public boolean canHandle(String token, int lineNumber, String fileName, ProgramArtifact artifact, TokenInfo tokenInfo) {
        // Use deterministic TokenInfo to check if this is a parameter token
        if (tokenInfo == null) {
            return false;
        }
        
        // Handle VARIABLE type tokens that are parameters
        // Parameters are marked as VARIABLE type in the procedure's scope
        // Also check if the scope is not "global" (parameters are in procedure scope)
        boolean isVariableType = tokenInfo.tokenType() == org.evochora.compiler.frontend.semantics.Symbol.Type.VARIABLE;
        boolean isInProcedureScope = !"global".equalsIgnoreCase(tokenInfo.scope());
        
        return isVariableType && isInProcedureScope;
    }
    
    @Override
    public TokenAnalysisResult analyze(String token, int lineNumber, ProgramArtifact artifact, TokenInfo tokenInfo, RawOrganismState o) {
        // Use the scope from TokenInfo directly - no need to scan backwards!
        String procName = tokenInfo.scope();
        if ("global".equalsIgnoreCase(procName)) {
            return null; // Not a procedure parameter
        }
        
        // Get the procedure's parameter names
        List<String> paramNames = artifact.procNameToParamNames().get(procName.toUpperCase());
        if (paramNames == null) {
            return null; // Procedure has no parameters
        }
        
        // Check if this token is a parameter name
        int paramIndex = -1;
        for (int i = 0; i < paramNames.size(); i++) {
            if (paramNames.get(i).equalsIgnoreCase(token)) {
                paramIndex = i;
                break;
            }
        }
        
        if (paramIndex == -1) {
            return null; // Not a parameter name
        }
        
        // Get the current value of this parameter by resolving the binding chain
        if (o != null && o.callStack() != null && !o.callStack().isEmpty()) {
            // Resolve the binding chain to find the actual register this parameter is bound to
            int finalRegId = resolveBindingChain(o.callStack(), paramIndex);
            if (finalRegId >= 0) {
                // Get the value from the final register
                Object value = getRegisterValue(finalRegId, o);
                if (value != null) {
                    String canonicalRegName = idToCanonicalName(finalRegId);
                    // Show only the physical register binding, not the parameter name
                    // Note: Frontend will add brackets, so don't include them here
                    String annotationText = String.format("%s=%s", canonicalRegName, formatValue(value));
                    return new TokenAnalysisResult(token, TokenType.PARAMETER_NAME, annotationText, "param");
                }
            }
        }
        
        // No value found - no annotation
        return null;
    }
    
    // Removed findCurrentProcedureName method - no longer needed!
    // We now use the scope directly from TokenInfo which is much simpler and more reliable.
    
    /**
     * Resolves the binding chain through the call stack to find the actual register.
     * Walks through FPR bindings to find the final DR/PR register.
     */
    private int resolveBindingChain(Deque<SerializableProcFrame> callStack, int initialFprIndex) {
        List<SerializableProcFrame> frames = new java.util.ArrayList<>(callStack);
        int currentRegId = Instruction.FPR_BASE + initialFprIndex;
        
        for (SerializableProcFrame frame : frames) {
            Integer mappedId = frame.fprBindings().get(currentRegId);
            if (mappedId != null) {
                currentRegId = mappedId;
                if (currentRegId < Instruction.FPR_BASE) {
                    return currentRegId; // Found a DR or PR
                }
            } else {
                // End of chain
                break;
            }
        }
        return currentRegId;
    }
    
    /**
     * Converts a register ID to its canonical name.
     */
    private String idToCanonicalName(int regId) {
        if (regId >= Instruction.FPR_BASE) return "%FPR" + (regId - Instruction.FPR_BASE);
        if (regId >= Instruction.PR_BASE) return "%PR" + (regId - Instruction.PR_BASE);
        if (regId >= 0) return "%DR" + regId;
        return "INVALID";
    }
    
    /**
     * Gets the runtime value of a register by ID.
     */
    private Object getRegisterValue(int regId, RawOrganismState o) {
        if (regId >= Instruction.FPR_BASE) {
            int index = regId - Instruction.FPR_BASE;
            return (o.fprs() != null && index < o.fprs().size()) ? o.fprs().get(index) : null;
        }
        if (regId >= Instruction.PR_BASE) {
            int index = regId - Instruction.PR_BASE;
            return (o.prs() != null && index < o.prs().size()) ? o.prs().get(index) : null;
        }
        if (regId >= 0 && o.drs() != null && regId < o.drs().size()) {
            return o.drs().get(regId);
        }
        return null;
    }
    
    /**
     * Formats a parameter value for display.
     * Handles molecules (TYPE:VALUE) and vectors (x|y|...).
     * Note: The register name is added by the caller, so we only format the value part.
     */
    private String formatValue(Object value) {
        if (value == null) return "null";
        
        if (value instanceof Integer i) {
            // Handle molecule values (TYPE:VALUE format)
            org.evochora.runtime.model.Molecule m = org.evochora.runtime.model.Molecule.fromInt(i);
            return String.format("%s:%d", typeIdToName(m.type()), m.toScalarValue());
        } else if (value instanceof int[] v) {
            // Handle vector values (x|y|... format, no brackets)
            return formatVector(v);
        } else if (value instanceof java.util.List<?> list) {
            // Handle vectors stored as List<Integer> after JSON deserialization
            return formatListAsVector(list);
        }
        
        return value.toString();
    }
    
    /**
     * Converts a type ID to its name.
     * Uses the central MoleculeTypeUtils utility to avoid duplication.
     */
    private String typeIdToName(int typeId) {
        return org.evochora.server.indexer.MoleculeTypeUtils.typeIdToName(typeId);
    }
    
    /**
     * Formats a vector as x|y|... (no brackets).
     */
    private String formatVector(int[] vector) {
        if (vector == null) return "";
        return java.util.Arrays.stream(vector).mapToObj(String::valueOf).collect(java.util.stream.Collectors.joining("|"));
    }
    
    /**
     * Formats a list as x|y|... (no brackets) for JSON deserialized vectors.
     */
    private String formatListAsVector(java.util.List<?> list) {
        if (list == null) return "";
        return list.stream().map(String::valueOf).collect(java.util.stream.Collectors.joining("|"));
    }
}

File: src/main/resources/web/visualizer/js/annotator/handlers/ParameterTokenHandler.js
/**
 * Handles the annotation of tokens that are procedure parameter names.
 * Shows the complete binding chain from source register to current FPR and its value.
 * Format: [%DR0→%FPR1→%FPR0=Value] where the chain shows data flow from source (DR/PR)
 * to intermediate bindings to the current parameter register (FPR), and Value is always
 * read from the current FPR (which may differ from the source if modified in the procedure).
 * Works anywhere in procedure bodies, not just on .PROC directive lines.
 * Resolves parameter bindings by walking the call stack to find the complete chain.
 */
class ParameterTokenHandler {
    /**
     * Determines if this handler can process the given token.
     * It handles tokens identified as 'VARIABLE' type that are in a procedure scope (not global).
     *
     * @param {string} tokenText The text of the token.
     * @param {object} tokenInfo Metadata about the token from the compiler.
     * @returns {boolean} True if the token is a 'VARIABLE' type in a procedure scope, false otherwise.
     */
    canHandle(tokenText, tokenInfo) {
        if (!tokenInfo) {
            return false;
        }
        
        const isVariableType = tokenInfo.tokenType === 'VARIABLE';
        const isInProcedureScope = tokenInfo.scope && tokenInfo.scope.toUpperCase() !== 'GLOBAL';
        
        return isVariableType && isInProcedureScope;
    }

    /**
     * Analyzes the parameter token to create an annotation with its binding chain and current value.
     * Resolves the complete binding chain through the call stack (from source DR/PR to current FPR),
     * then reads the value from the current FPR (not from the source, as FPR values may be modified).
     * The binding chain is displayed reversed: from source to target (e.g., %DR0→%FPR1→%FPR0).
     *
     * @param {string} tokenText The text of the token (the parameter name).
     * @param {object} tokenInfo Metadata about the token.
     * @param {object} organismState The current state of the organism, containing the call stack.
     * @param {object} artifact The program artifact containing `procNameToParamNames` and `callSiteBindings`.
     * @returns {object} An annotation object `{ annotationText, kind }`.
     * @throws {Error} If required data (procedure name, parameter list, call stack, binding chain) is missing or invalid.
     */
    analyze(tokenText, tokenInfo, organismState, artifact) {
        if (!organismState || !organismState.callStack || !Array.isArray(organismState.callStack)) {
            throw new Error(`Cannot annotate parameter "${tokenText}": organismState.callStack is missing or invalid.`);
        }

        if (!artifact || !artifact.procNameToParamNames || typeof artifact.procNameToParamNames !== 'object') {
            throw new Error(`Cannot annotate parameter "${tokenText}": artifact.procNameToParamNames is missing or invalid.`);
        }

        // Get procedure name from token scope
        const procName = tokenInfo.scope;
        if (!procName || procName.toUpperCase() === 'GLOBAL') {
            throw new Error(`Cannot annotate parameter "${tokenText}": token scope is not a procedure name (scope: "${procName}").`);
        }

        // Get the procedure's parameter information
        // Structure: procNameToParamNames["PROC1"] = { params: [{name: "PARAM1", type: "REF"}, {name: "PARAM2", type: "VAL"}] }
        const paramNamesEntry = artifact.procNameToParamNames[procName.toUpperCase()];
        if (!paramNamesEntry || typeof paramNamesEntry !== 'object') {
            throw new Error(`Cannot annotate parameter "${tokenText}": procedure "${procName}" not found in procNameToParamNames.`);
        }

        if (!paramNamesEntry.params || !Array.isArray(paramNamesEntry.params)) {
            throw new Error(`Cannot annotate parameter "${tokenText}": procedure "${procName}" has no params array in parameter entry.`);
        }

        // Extract parameter names from ParamInfo array
        const params = paramNamesEntry.params;
        const paramNames = params.map(p => p.name);

        // Find the parameter index
        let paramIndex = -1;
        for (let i = 0; i < paramNames.length; i++) {
            if (paramNames[i] && paramNames[i].toUpperCase() === tokenText.toUpperCase()) {
                paramIndex = i;
                break;
            }
        }

        if (paramIndex === -1) {
            throw new Error(`Cannot annotate parameter "${tokenText}": not found in procedure "${procName}" parameter list.`);
        }

        // Resolve the binding chain through the call stack using artifact bindings
        // resolveBindingChainWithPath returns the complete path of register IDs in display order
        // Path format: [DR0, FPR1, FPR0] - from source DR to parameter FPR
        const bindingPath = AnnotationUtils.resolveBindingChainWithPath(paramIndex, organismState.callStack, artifact, organismState);

        if (bindingPath.length === 0) {
            throw new Error(`Cannot annotate parameter "${tokenText}": binding path is empty.`);
        }

        // Get the value from the LAST register in the chain (the currently valid FPR)
        // bindingPath ends with the parameter's FPR (e.g., [DR0, FPR1, FPR0])
        // This FPR holds the current value, which may differ from the source DR/PR if modified
        const currentRegId = bindingPath[bindingPath.length - 1];

        // Get the register value from the current FPR
        // getRegisterValueById now throws Error directly if register not found or invalid input
        const value = AnnotationUtils.getRegisterValueById(currentRegId, organismState);

        // Format the binding path: %DR0→%FPR1→%FPR0
        // The path is already in display order (source to target)
        const pathNames = bindingPath.map(regId => AnnotationUtils.formatRegisterName(regId));
        const pathDisplay = pathNames.join('→');

        // Format the value
        const formattedValue = ValueFormatter.format(value);

        // Always show the complete chain (even if it has only one element)
        // Format: [%DR0→%FPR1→%FPR0=Value] where Value is from FPR0 (current, last element)
        return {
            annotationText: `[${pathDisplay}=${formattedValue}]`,
            kind: 'param'
        };
    }
}

window.ParameterTokenHandler = ParameterTokenHandler;

